factcast-snapshotcache-mongodb/src/main/java/org/factcast/core/snap/mongo/MongoDbSnapshotCache.java

```diff
@@ -0,0 +1,267 @@
+    // Warmup the gridFS bucket by doing a dummy upload and delete (Its forbidden later in txns)
+    byte[] tiny = new byte[] {0};
+    ObjectId tmp = gridFSBucket.uploadFromStream("_warmup_", new ByteArrayInputStream(tiny));
+    gridFSBucket.delete(tmp); // cleanup
+
+    // Schedule cleanup task
+    scheduleCleanupTask();
+  }
+
+  @Override
+  @SneakyThrows
+  public @NonNull Optional<SnapshotData> find(@NonNull SnapshotIdentifier id) {
+    Document query = new Document(IDENTIFIER_FIELD, getDocumentIdentifier(id));
+    Document result = collection.find(query).first();
+
+    if (result == null) {
+      return Optional.empty();
+    }
+
+    // Field ID of the binary in GridFS
+    ObjectId fileId = result.getObjectId(FILE_ID_FIELD);
+
+    byte[] bytes;
+    // Stream the GridFS file to a destination (here, a file on disk)
+    try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+      gridFSBucket.downloadToStream(fileId, out);
+      bytes = out.toByteArray();
+    }
+
+    String serializerId = result.getString(SNAPSHOT_SERIALIZER_ID_FIELD);
+    UUID lastFactId = UUID.fromString(result.getString(LAST_FACT_ID_FIELD));
+
+    tryUpdateExpirationDateAsync(id);
+
+    return Optional.of(new SnapshotData(bytes, SnapshotSerializerId.of(serializerId), lastFactId));
+  }
+
+  @Override
+  public void store(@NonNull SnapshotIdentifier id, @NonNull SnapshotData snapshot) {
+    String documentIdentifier = getDocumentIdentifier(id);
+    Document doc =
+        new Document(IDENTIFIER_FIELD, documentIdentifier)
+            .append(PROJECTION_CLASS_FIELD, id.projectionClass().getName())
+            .append(SNAPSHOT_SERIALIZER_ID_FIELD, snapshot.snapshotSerializerId().name())
+            .append(LAST_FACT_ID_FIELD, snapshot.lastFactId().toString())
+            .append(
+                EXPIRE_AT_FIELD,
+                Instant.now().plus(properties.getDeleteSnapshotStaleForDays(), ChronoUnit.DAYS));
+
+    UUID aggregateId = id.aggregateId();
+    if (aggregateId != null) {
+      doc.append(AGGREGATE_ID_FIELD, aggregateId.toString());
+    }
+
+    Document query = new Document(IDENTIFIER_FIELD, documentIdentifier);
+    String fileName = "filename_" + documentIdentifier;
+
+    Document result = collection.find(query).first();
+
+    try (ClientSession session = mongoClient.startSession()) {
+      session.withTransaction(
+          () -> {
+            if (result != null) {
+              // Delete old snapshot file
+              ObjectId oldFileId = result.getObjectId(FILE_ID_FIELD);
+              gridFSBucket.delete(session, oldFileId);
+            }
+
+            ObjectId fileId;
+            try (InputStream in = new ByteArrayInputStream(snapshot.serializedProjection())) {
+              fileId = gridFSBucket.uploadFromStream(session, fileName, in);
+            } catch (IOException e) {
+              log.error("Error uploading snapshot to GridFS for id: {}", id, e);
+              throw new RuntimeException(e);
```

Comment:
can we just log an error here? Nobody would catch a generic RuntimeExceptionm, so we would break any process here unnecessary (the process could complete without storing the snapshot)